src/django_components/html.py

```diff
@@ -0,0 +1,98 @@
+import re
+from typing import Dict, Optional, Protocol, Union, cast
+
+from selectolax.lexbor import LexborHTMLParser, LexborNode
+
+
+def parse_node(html: str) -> LexborNode:
+    parser = LexborHTMLParser(html)
+    # NOTE: The parser automatically places <style> tags inside <head>
+    # while <script> tags are inside <body>.
+    return parser.body.child or parser.head.child  # type: ignore[union-attr, return-value]
+
+
+# NOTE: While Selectolax offers way to insert a node before or after
+# a current one, it doesn't allow to insert one node into another.
+# See https://github.com/rushter/selectolax/issues/126
+def html_insert_before_end(html: str, insert_html: str) -> str:
+    regex = re.compile(r"<\/\w+>$")
+    return regex.sub(
+        lambda m: insert_html + m[0],
+        html.strip(),
+    )
+
+
+def insert_before_end(node: LexborNode, insert: Union[LexborNode, str]) -> None:
+    the_insert = insert.html or "" if isinstance(insert, LexborNode) else insert
+    new_node_html = html_insert_before_end(node.html or "", the_insert)
+    new_node = parse_node(new_node_html)
+    node.replace_with(new_node)  # type: ignore[arg-type]
+    node.insert_before(new_node)  # type: ignore[arg-type]
+
+
+HTML_ROOT_TAGS_REGEX = re.compile(
+    r"<!doctype|<html|</html|<head|</head|<body|</body",
+    re.IGNORECASE,
+)
+
+HTML_ROOT_TAGS_REVERSE_REGEX = re.compile(
+    r"<doctype_|<html_|</html_|<head_|</head_|<body_|</body_",
+    re.IGNORECASE,
+)
+
+
+html_tag_escapes: Dict[str, str] = {
+    "<!doctype": "<doctype_",
+    "<html": "<html_",
+    "</html": "</html_",
+    "<head": "<head_",
+    "</head": "</head_",
+    "<body": "<body_",
+    "</body": "</body_",
+}
+html_tag_escapes_reverse = {val: key for key, val in html_tag_escapes.items()}
+
+
+class TransformHtmlCallback(Protocol):
+    def __call__(self, head: Optional[LexborNode], body: Optional[LexborNode]) -> None: ...  # noqa: #704
+
+
+def transform_html_document(html: str, transform: TransformHtmlCallback) -> str:
+    # Escape <!doctype>, <html>, <head>, and <body> tags, because Selectolax treats them
+    # specially, which makes it impossible to edit them.
```

Comment:
> Since we are already detecting the head and body end tags, couldn't we insert our HTML directly with a string replace, and not need selectolax at all?

Unfortunately inserting with string replace its not that simple. Consider this example:

```html
<!-- <body> is bla bla bla -->
<html>
  <body>
      <script>
        console.log("Searching for <body>");
        document.querySelector("body");
      </script>
  </body>
</html>
```

With a naive regex, we wouldn't be able to sensibly select only the correct `<body>` without replacing also also all entries of `<body>` in comments or scripts.

In my case I could use regex, because I'm not inserting anything extra, I'm just temporarily renaming a string and then renaming it back. And even then I'm not 100% if there won't be some edge case.

> So in the first usage of this new library we have to work around very basic limitations... Doesn't bode well for the future I think. Are you still confident this is the right choice?

Yeah, the Python wrapper is not that great, but the underlying C libraries are solid. [See this Reddit thread](https://www.reddit.com/r/Python/comments/vu9fgz/comment/ifg3pj1/?utm_source=share&utm_medium=web3x&utm_name=web3xcss&utm_term=1&utm_content=share_button) (which links to [this performance comparison](https://gist.github.com/MercuryRising/4061368?permalink_comment_id=3004245#gistcomment-3004245)).

Here are my own tests:

Selectolax is 4-4.5x as fast as lxml when it comes to pure parsing:
```
==== Total trials: 5000 =====
LXML total time: 12.8 s
LEXBOR total time: 2.8 s
```

And about 3x as fast as lxml when it comes to mutating HTML:

```
==== Total trials: 5000 =====
LEXBOR - Total time: 6.2 s
LXML - Total time: 19.2 s
```

But the HTML mutation with my workarounds is actually slower than lxml:

```
==== Total trials: 5000 =====
LEXBOR - Total time: 20.0 s
LXML - Total time: 16.9 s
```

So if we got the issues resolved upstream, we'd get a speed-up of 3-4x. At the same time, implementing the features like scoped CSS and whatnot will overall require the HTML to be parsed 2-3x. So in the end we should end up with comparable performance than if we parsed the HTML once with lxml.

The upstream issue consists of 2 parts, and I've reported it in https://github.com/rushter/selectolax/issues/128:

1. API for inserting one node into another
   - Here, [Lexbor DOES implement that](https://github.com/lexbor/lexbor/blob/72236d31da1a4c3d83c621446dcf4f70345fda05/source/lexbor/dom/interfaces/node.c#L385). It's just missing from the Selectolax wrapper, but adding it should be as simple as [copying this function and updating function calls](https://github.com/rushter/selectolax/blob/4370cddf04a6ed3e8cad97f2d588bb5f6c854511/selectolax/lexbor/node.pxi#L556).

2. Making sure methods like `Node.replace_with()` or `Node.insert_child()` work with `<head>` and `<body>` tags.
   - This is an unknown for me, and I will need to see what the maintainers of Selectolax have to say.
      - Tho I can imagine that it could turn out that `Node.replace_with()` might not work with head / body, while methods like `Node.insert_child()` would work. Because `Node.insert_after/before()` work fine.

Reply:
Note: [I've added `Node.insert_child()` upstream](https://github.com/rushter/selectolax/pull/130) so once its released, we should be able to remove `transform_html_document`

Reply:
Sorry for spam, but good news, Selectolax fix was released, and in my local code (not in this PR) I managed to replace the workaround code with the code below. 

```py
tree = LexborHTMLParser(html_content)

# Parse the head and body snippets into elements
_, head_snippet_elements = parse_multiroot_html(head_snippet)
_, body_snippet_elements = parse_multiroot_html(body_snippet)

# Append each element in the head snippet to the <head> tag
for element in head_snippet_elements:
    tree.head.insert_child(element)

# Append each element in the body snippet to the <body> tag
for element in body_snippet_elements:
    tree.body.insert_child(element)

# Return the modified HTML content as a string
return tree.html
```

And this is still 3x faster than lxml:

```
==== Total trials: 5000 =====
LEXBOR - Total time: 8.5
LXML - Total time: 24.3
```

Reply:
Awesome!